keyext.proofmanagement/src/main/java/org/key_project/proofmanagement/check/dependency/ContractAppCollector.java

```diff
@@ -0,0 +1,216 @@
+        String axiomName = tacletApp.getRuleName().replace('_', ' ');
+        Set<ClassAxiom> axioms = specRepo.getClassAxioms(classType).toSet();
+        List<ClassAxiom> axiomList = axioms.stream()
+                .filter(a -> a.getName().equals(axiomName))
+                .collect(Collectors.toList());
+        /*
+         * the assertions below always hold for current KeY implementation, where
+         * only one model method with same name is allowed (no overloading)
+         */
+        assert axiomList.size() == 1;
+        ClassAxiom axiom = axiomList.get(0);
+
+        // found axiom always is a contract axiom, target always a program (model) method!
+        assert axiom instanceof ContractAxiom;
+        ContractAxiom contractAxiom = (ContractAxiom) axiom;
+        IObserverFunction obs = contractAxiom.getTarget();
+        assert obs instanceof ProgramMethod;
+        ProgramMethod modelMethod = (ProgramMethod) obs;
+        assert modelMethod.getMethodDeclaration().isModel();
+
+        // find the searched contract with specRepo, class type, and program method
+        Set<Contract> contractsSet = specRepo.getContracts(classType, modelMethod).toSet();
+        List<Contract> contractList = new ArrayList<>(contractsSet);
+
+        // TODO: this is not the case if multiple contracts are combined with also keyword,
+        // however, in this case the contract axioms are completely broken anyway
+        assert contractList.size() == 1;
+        Contract contract = contractList.get(0);
+
+        // model method contract application is always termination sensitive!
+        result.putIfAbsent(contract.getName(), TERMINATION_SENSITIVE);
+    }
+
+    /**
+     * Extracts the contracts from a builtin rule. Note that these may be multiple contracts,
+     * since KeY sometimes combines contracts!
+     *
+     * @param biApp the builtin rule node to extract the contracts from
+     */
+    private void extractContractsFromBuiltin(BuiltInAppIntermediate biApp) {
+        // The string may still contain multiple contracts, syntax: contract1#contract2#...
+        // split and add all
+        SpecificationRepository specRepo = proof.getServices().getSpecificationRepository();
+        Contract c = specRepo.getContractByName(biApp.getContract());
+        ImmutableSet<Contract> contracts = specRepo.splitContract(c);
+
+        // load information about the modality under which the contract was applied
+        Modality modality = Modality.getModality(biApp.getModality());
+        DependencyGraph.EdgeType edgeType;
+        if (modality == null) {
+            // in default case (e.g. legacy proofs without saved modality information)
+            // we assume diamond modality but print a warning
+            logger.print(LogLevel.WARNING, "No saved modality information was found!" +
+                " Assuming \"diamond\" (incomplete for box contracts)!");
```

Comment:
Shouldn't we assume box/termination-insensitive to be sound?

Reply:
No. Inside a box modality, it is always sound to apply any (matching) contract. This means that such a contract application is ignored for the dependency check (not included into any strongly connected component). However, since potentially the contract could have been applied from a diamond modality (we do not know), we need to be conservative and assume that the modality was a diamond. This way the contract application (edge) can be included in the strongly connected components and needs to have a termination measure to be legal. This might mark some contract applications as illegal and generate false positives (cycles marked as illegal even though they are not), which then could be fixed easily by adding the modality information to the proof manually (with a text editor).

I think that most of the time the contracts and modalities in use should be termination sensitive, so it should work out of the box in the majority of cases.